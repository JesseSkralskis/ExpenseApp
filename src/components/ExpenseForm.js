import React, { Component } from 'react';
//needed to make react-dates work
import "react-dates/initialize";
//named export from the react dates library
import uuid from 'uuid';
import database from '../firebase/firebase'
import { SingleDatePicker } from 'react-dates';

//for docu go to https://github.com/airbnb/react-dates
//tools used to manipulate date:
//1.moment 2. air b and b react-dates
import moment from 'moment';
 //api for this javascript way is dog shit so its a no go
// const date = new Date();
//so we use moment
//for documentation on patterns go to
//https://momentjs.com/docs/#/displaying/
const now = moment();




export default class ExpenseForm extends Component {
                 //we need local compomnent state to track all of the changes in these inputs
                 //only when the user submits the form will we actually do something with the info
  constructor(props) {
    super(props);
   
    this.state = {
      //because we want to populate the fields if the user has chosen to edit
      //we use turnerary logic that checks if the expense object has been passed from the parent
      //if it has it means the user is editing so will populate
      
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
    //divide by 100 so the field holds the proper decimal situation
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      //to poopulate the correct date we call moment with the created at paramater
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: ""
    };
  }
             
                 //create an event handler function
                 handleDescriptionChange = e => {
                   const description = e.target.value;
                   this.setState(() => ({
                     description
                   }));
                 };
                 handleNoteChange = e => {
                   const note = e.target.value;
                   this.setState(() => ({
                     note
                   }));
                 };

    onAmountChange = e => {
                   const amount = e.target.value;
                   //function that adds restrictions to inputs by checking if input value matches
                   //regex expression
                   if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
                     this.setState(() => ({
                       amount
                     }));
                   } else {
                   }
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({
            calenderFocused: focused
        }));
    }
    
    
    
    
  handleDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  }
  
  handleOnSubmit = (e) => {
    //on submit need to do this so page doesnt do full refresh handles with just javascript
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'you need to fill out the description or amount field'
      }))
      
    } else {
      this.setState(() => ({
        error: ""
      }));
      //calling the function that was passed from parent with the object from all the state from this child component
      this.props.onSubmit({
        
        description: this.state.description,
        //because amount is in string form we must parse it also multiply by 100 
        //so we can negate that it is in cents
        amount: parseFloat(this.state.amount, 10) * 100,
        //because the state is set to moment we need to parse this back to its og form
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
        

      })
    }
    
  }
                 render() {
                   return (
                     <form className="form" onSubmit={this.handleOnSubmit}>
                       {this.state.error && (
                         <p className="form__error">{this.state.error}</p>
                       )}
                       <input
                         className="text-input"
                         type="text"
                         placeholder="description"
                         //automatically places the cursor on this input
                         autoFocus
                         value={this.state.description}
                         onChange={this.handleDescriptionChange}
                       />
                       {/* becuse type number allows many decimalplaces we must use text */}
                       <input
                         className="text-input"
                         type="text"
                         placeholder="amount"
                         value={this.state.amount}
                         onChange={this.onAmountChange}
                       />

                       <SingleDatePicker
                         date={this.state.createdAt}
                         //similr to other event handlers but difference is
                         //this one is called by our third party library
                         onDateChange={this.handleDateChange}
                         focused={this.state.calenderFocused}
                         onFocusChange={this.onFocusChange}
                         //custumization
                         //docu https://github.com/airbnb/react-dates
                         numberOfMonths={1}
                         isOutsideRange={() => {
                           return false;
                         }}
                       />

                       <textarea
                         className="text-area"
                         placeholder="add a note for your expense (optional)"
                         value={this.state.note}
                         onChange={this.handleNoteChange}
                       />
                       <div>
                         <button value="ballsack" className="buttons">{window.location.href.indexOf('edit')> -1 ?"Edit Expense": "Add expense"}</button>
                       </div>
                     </form>
                   );
                 }
               }
