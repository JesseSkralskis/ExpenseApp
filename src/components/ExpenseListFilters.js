import React from 'react'
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate,setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from 'react-dates';
import uuid from 'uuid';
 
 export class ExpenseListFilters extends React.Component{
    
    state = {
        calenderFocused: null
    };
   
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({
            calenderFocused 
        }))
        
    }
    onDatesChange = ({startDate, endDate}) => {
      this.props.setStartDate(startDate);
      this.props.setEndDate(endDate);
    }
   
   
   onTextChange = (e) => {
     this.props.setTextFilter(e.target.value);
   }
   onSortByChange = (e) => {
     if (e.target.value === "date") {
       this.props.sortByDate();
     } else if (e.target.value === "amount") {
       this.props.sortByAmount();
     }
   }

   
    render() {
       
      return (
        <div className="content-container">
          <div className="input-group">
            <div className="input-group__item">
              {" "}
              <input
                className="text-input"
                placeholder="search expenses"
                type="text"
                value={this.props.filters.text}
                onChange={this.onTextChange}
              />
            </div>
            <div className="input-group__item">
              <select className="select"
                value={this.props.filters.sortBy}
                onChange={this.onSortByChange}
              >
                <option value="date">Date</option>
                <option value="amount">Amount</option>
              </select>
            </div>
            <div className="input-group__item">
              <DateRangePicker
                startDate={this.props.filters.startDate}
                startDateId={"abcd5678"}
                endDate={this.props.filters.endDate}
                endDateId={"abc1235"}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calenderFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
                showClearDates={true}
              />
            </div>
          </div>
        </div>
      );
    }

 }

const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (value) => dispatch(setTextFilter(value)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
    
   
 })

 
const mapStateToProps = (state) => {
    return {
         filters: state.filters
     }
 }

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);
