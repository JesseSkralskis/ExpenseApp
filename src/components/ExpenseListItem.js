import React from 'react'
import { NavLink } from "react-router-dom";
import moment from 'moment';
import numeral from 'numeral';

 



//this takes the props from expenseList (destructerted) and placesd them as needed
const ExpenseListItem = ({  id, description, amount, createdAt }) => (
  <div>
    <NavLink to={`/edit/${id}`}>
      {" "}
      <h3>{description}</h3>
    </NavLink>
    <p>
      {numeral(amount/100).format('$0,0.00')}
      -
      {moment(createdAt).format('MMMM Do, YYYY ')}
    </p>
    
  </div>
);
   //when we add connect we get dispatch as a prop even if we do no mapToProps function
    export default ExpenseListItem;