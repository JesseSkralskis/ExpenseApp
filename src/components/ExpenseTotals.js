import React from 'react';
import selectExpenses from '../selectors/expenses';
import selectExpenseTotal from '../selectors/expenses-total'
import { connect } from "react-redux";
import numeral from "numeral";

export const ExpenseTotals = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formattedTotal = numeral(expensesTotal/100).format('$0,0.00');

    return (
        <div>
            <h1>Viewing {expensesCount} {expenseWord} totalling {formattedTotal}</h1>
        </div>
    )
}
// here I use the selector to get all the visible expenses
//nthen I pass the props objects up 
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
   
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpenseTotal(visibleExpenses)
        
    }

}

export default connect(mapStateToProps)(ExpenseTotals);