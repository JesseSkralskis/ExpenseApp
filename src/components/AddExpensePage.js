import React from "react";

import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from '../actions/expenses'
//this component which is the parent to the expense form passes a prop down to its child
//that is a function that will take the expense info that a user fills out
//and then dispatch the adExpense action on that data which will change the redux store to add
//to the array of expenses

export class AddExpensePage extends React.Component {
  
  onSubmit = (expense) => {
   
    //because of mapDispatchToProps we can now abtract our dispatcher functions away from the component the code which will make it testable
    // props.dispatch(addExpense(expense));
    this.props.startAddExpense(expense);
    //react router has props handed to it in this case we use push to redirect
   this.props.history.push("/");
  };
  render() {
    
    
    return (
      <div>
        <h1>Add Expense</h1>

        <ExpenseForm
          //passing this prop down to the child which is the form
          //now this function is no longer "inline" meaning it doesnt get called every render
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}


// connect seccond argument 
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: ( expense ) => dispatch(startAddExpense( expense ))
});


export default connect(undefined , mapDispatchToProps)(AddExpensePage);
