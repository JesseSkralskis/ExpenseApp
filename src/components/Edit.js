import React from "react";
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from "../actions/expenses";
//refactor to pull out in line functions, turn them into methods so they dont have to rerender
// set up mapDispatchToProps 2 things

export class Edit extends React.Component{
  
  onSubmit = (expense) => {
   console.log(expense)
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
    
  }

  onRemove = () => {
   this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  }
  render() {
    return (
      <div>
        {/* in this instance of expense form we need to populate the fields so we
      pass some props to the child the expense match
      and what should happen when they push the button */}
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >
          Remove
        </button>
      </div>
    );
  }
}



//we want to be able to pass the expense that matches the id that the user clicked on
//so using the connect state and props we can search through the expenses array in the store
//returning the 1 expense that matches the ids
//then that expense object i passed up to the component through props
const mapToProps = (state,props) => {
 
  return {
    //find allows us to look through an array looking for a single item
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id
    })
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense( id , expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
})
  

  


export default connect(mapToProps,mapDispatchToProps)(Edit);


 