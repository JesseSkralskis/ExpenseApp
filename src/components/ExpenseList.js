import React from 'react'
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'


// 1. regular unconnected component

//important to remember if you are connected to the store your component is reactive and will rerender
//now wecan use the state of our store as props in the componenent
 export const ExpenseList=(props) =>{
    return (
        <div>
            {/* for test we creat turneary logic to test if there are no expenses */}
            {
                props.expenses.length === 0 ? (
                    <p>No expenses</p>
                ) : (
                     props.expenses.map((expense) => {
              
                //spread out al the things on expense which is all our propertys we are rendering in ListItem
                //now ExpenseList Item has access to every individual instance with all the pairs 
                return <ExpenseListItem key={expense.id}{...expense}/>
           })
               
                )
            }

           
            
        </div>
    )
}

//create higher order component
//using there react-redux api thats why expense list is outside of brackests
//give the component a subset from the store
//passes state into connect
// 2. common practice create a mapping function
const mapStateProps = (state) => {
    return {
        //to get our expenses we now call our selector filter with the needed state
        expenses: selectExpenses(state.expenses, state.filters)
    }
}
//.3 export ties it all together
//saying connect expense list with the props from the higher order function
export default connect(mapStateProps)(ExpenseList);



