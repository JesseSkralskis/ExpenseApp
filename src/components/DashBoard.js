import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseTotals from './ExpenseTotals';
import ExpenseListFilters from '../components/ExpenseListFilters';


 
//need to import the named export from redux-react to be able to get the store
//this is the parent of expense list



const DashBoard = () => (
   
    <div>
        <ExpenseTotals/>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)


 
export default DashBoard;