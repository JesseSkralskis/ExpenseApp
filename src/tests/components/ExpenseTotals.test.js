import React from 'react';
import { ExpenseTotals } from '../../components/ExpenseTotals';
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json';

test('should render the correct message with one value', () => {
    const wrapper = shallow(<ExpenseTotals expensesCount={1} expensesTotal={235} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render the correct message with multiple values", () => {
  const wrapper = shallow(
    <ExpenseTotals expensesCount={23} expensesTotal={278790} />
  );
  expect(toJSON(wrapper)).toMatchSnapshot();
});

