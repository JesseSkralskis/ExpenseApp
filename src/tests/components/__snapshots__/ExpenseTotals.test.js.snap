// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the correct message with multiple values 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $2,787.90
  </h1>
</div>
`;

exports[`should render the correct message with one value 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;
